File: 后端篇/disc/第7章/7.1/server/node/controllers/api/histories.js
// JS：controllers\api\histories.js
"use strict"
import model from "../../models/index.js"

export default {
  /**
   * GET /histories：查询个人的历史记录
   * @param {*} ctx 
   * @param {*} next 
   * @returns {obejct} {"errMsg":"ok","data":[{}]}
   */
  "GET ": async (ctx, next) => {
    const { openid } = ctx.request.query
      , res = { errMsg: "ok" }
      , db = await model.connect()
    res.data = await model.execute(db, "select * from history where openid=@openid order by date_created desc limit 10", { openid })
    ctx.status = 200
    ctx.body = res
  },

  /**
   * POST /histories：新增一条历史记录
   * @returns {*} {"errMsg":"ok","data":newid}
   */
  "POST ": async (ctx, next) => {
    const { openid, user_score, system_score } = ctx.request.body
      , activity = model.history({ openid, user_score, system_score })
      
    await activity.save() // 即使成功返回也是undefined
    const res = { errMsg: activity.identity() > 0 ? "ok" : "", data: activity.identity() }
    ctx.status = 200
    ctx.body = res
  },

  /**
   * DELETE /histories/{history_id}：依据history_id删除一条记录
   * @param {*} ctx 
   * @param {*} next 
   * @returns {object} {"errMsg":"ok"}
   */
  "DELETE /:history_id": async (ctx, next) => {
    let { history_id: id } = ctx.params
    const { openid } = ctx.request.query
      , db = await model.connect()
    id = parseInt(id)
    /**
     * 返回的数据示例
     * {affectedRows: 1
        changedRows: 0
        fieldCount: 0
        insertId: 0
        message: ""}
     */
    // 只用一个id就足够了，这里演示两个查询的使用
    const res = await model.execute(db, "delete from history where id=@id && openid=@openid limit 1", { id, openid })
    ctx.status = 200
    ctx.body = {
      errMsg: res.affectedRows > 0 ? "ok" : "err"
      , data: res.affectedRows
    }
  }
}

File: 番外篇/disc/第2章/server/go/controllers/api/histories.go
// GO：controllers\api\histories.go
package api

import (
	"fmt"
	"local/models"
	"github.com/kataras/iris/v12"
	"xorm.io/xorm"
)

// 基地址：/api/histories
type HistoryController struct {
	Ctx iris.Context // HTTP请求上下文对象，由iris自动绑定
	Db  *xorm.Engine
}

// GET：/api/histories
// curl -X GET http://localhost:3000/api/histories?openid=o0_L54sDkpKo2TmuxFIwMqM7vQcU
func (c *HistoryController) Get() models.Result {
	openid := c.Ctx.URLParam("openid")
	list := make([]models.History, 0)
	c.Db.Where("openid = ?", openid).Limit(10, 0).Find(&list)
	return models.Result{Data: list, ErrMsg: "ok"}
}

// POST：/api/histories
// curl -X POST -H "Content-Type: application/json" -d '{"openid":"o0_L54sDkpKo2TmuxFIwMqM7vQcU","system_score":0,"user_score":1}' http://localhost:3000/api/histories
func (c *HistoryController) Post() models.Result {
	var res = models.Result{}
	activity := new(models.History)

	if err := c.Ctx.ReadJSON(&activity); err != nil {
		res.ErrMsg = "参数错误"
		return res
	}

	if affected, err := c.Db.InsertOne(activity); err != nil {
		fmt.Printf("%v\n", err)
		res.ErrMsg = err.Error()
		return res
	} else if affected > 0 {
		fmt.Printf("affected:%v\n", affected)
		res.ErrMsg = "ok"
		res.Data = activity.Id
	}

	return res
}

// DELETE：/api/histories/:id
// curl -X DELETE http://localhost:3000/api/histories/1?openid=o0_L54sDkpKo2TmuxFIwMqM7vQcU
func (c *HistoryController) DeleteBy(id int64) models.Result {
	openid := c.Ctx.URLParam("openid")
	res := models.Result{}
	if affected, _ := c.Db.ID(id).Where("openid = ?", openid).Delete(new(models.History)); affected > 0 {
		res.ErrMsg = "ok"
		res.Data = affected
	}

	return res
}
